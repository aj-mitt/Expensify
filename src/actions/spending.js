import uuid from 'uuid'
import database from '../firebase/firebase'

export const addSpending = (spending) => ({
  type: 'ADD_SPENDING',
  spending
})

export const startAddSpending = (spendingData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = spendingData
    const spending = {description, note, amount, createdAt}

    return database.ref('spending').push(spending).then((ref) => {
      dispatch(addSpending({
        id: ref.key,
        ...spending
      }))
    })
  }
}

export const removeSpending = ({id} = {}) => ({
  type: 'REMOVE_SPENDING',
  id
})

export const startRemoveSpending = ({id} = {}) => {
  return (dispatch) => {
    return database.ref(`spending/${id}`).remove().then(() => {
      dispatch(removeSpending({id}))
    })
  }
}

export const editSpending = (id, updates) => ({
  type: 'EDIT_SPENDING',
  id,
  updates
})

export const startEditSpending = (id, updates) => {
  return (dispatch) => {
    return database.ref(`spending/${id}`).update(updates).then(() => {
      dispatch(editSpending(id, updates))
    })
  }
}

export const setSpending = (spending) => ({
  type: 'SET_SPENDING',
  spending
})

export const startSetSpending = () => {
  return (dispatch) => {
    return database.ref('spending').once('value').then((snapshot) => {
      const spending = []

      snapshot.forEach((childSnapshot) => {
        spending.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
      dispatch(setSpending(spending))
    })
  }
}
