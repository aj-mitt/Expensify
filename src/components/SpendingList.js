import React from 'react'
import {connect} from 'react-redux'
import SpendingListItem from './SpendingListItem'
import selectSpending from '../selectors/spending'

export const SpendingList = (props) => (
  <div>
    {
      props.spending.length === 0 ? (
        <p> No Spending</p>
      ) : (
        props.spending.map((spending) => {
          return <SpendingListItem key={spending.id} {...spending}/>
        })
      )
    }
  </div>
)

const mapStateToProps = (state) => {
  return {
    spending: selectSpending(state.spending, state.filters)
  }
}

export default connect(mapStateToProps)(SpendingList)
