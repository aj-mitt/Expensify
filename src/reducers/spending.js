const spendingReducerDefaultState = []

export default (state = spendingReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_SPENDING':
      return [
        ...state,
        action.spending
      ]
    case 'REMOVE_SPENDING':
      return state.filter(({id}) => id !== action.id)
    case 'EDIT_SPENDING':
      return state.map((spending) => {
        if (spending.id === action.id) {
          return {
            ...spending,
            ...action.updates
          }
        } else {
          return spending
        }
      }
      )
    case 'SET_SPENDING':
      return action.spending
    default:
      return state
  }
}
