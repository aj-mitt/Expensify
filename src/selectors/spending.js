import moment from 'moment'

export default (spending, {text, sortBy, startDate, endDate}) => {
  return spending.filter((spending) => {

    const createdAtMomnet = moment(spending.createdAt)
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMomnet, 'day') : true
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMomnet, 'day') : true
    const textMatch = spending.description.toLowerCase().includes(text.toLowerCase())

    return startDateMatch && endDateMatch && textMatch
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1
    }
  })
}
