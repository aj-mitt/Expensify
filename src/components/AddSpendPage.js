import React from 'react'
import {connect} from 'react-redux'
import SpendingForm from './SpendingForm'
import {startAddSpending} from '../actions/spending'

export class AddSpendPage extends React.Component {
  onSubmit = (spending) => {
    this.props.startAddSpending(spending)
    this.props.history.push('/')
  }
  render() {
    return(
      <div>
      <h1>Add Spending</h1>
      <SpendingForm
        onSubmit={this.onSubmit}
      />
    </div>
      
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddSpending: (spending) => dispatch(startAddSpending(spending))
})

export default connect(undefined, mapDispatchToProps)(AddSpendPage)
