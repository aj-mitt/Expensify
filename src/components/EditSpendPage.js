import React from 'react'
import {connect} from 'react-redux'
import SpendingForm from './SpendingForm'
import { startEditSpending, startRemoveSpending } from '../actions/spending';

export class EditSpendPage extends React.Component {
  onSubmit = (spending) => {
    this.props.startEditSpending(this.props.spending.id, spending)
    this.props.history.push('/')
  }
  onRemove = () => {
    this.props.startRemoveSpending({id: this.props.spending.id})
    this.props.history.push('/')
  }
  render(){
    return (
      <div>
        <SpendingForm
          spending={this.props.spending}
          onSubmit= {this.onSubmit}
        />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
    spending: state.spending.find((spending) => spending.id === props.match.params.id)
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  startEditSpending: (id, spending) => dispatch(startEditSpending(id, spending)),
  startRemoveSpending: (data) => dispatch(startRemoveSpending(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditSpendPage)
