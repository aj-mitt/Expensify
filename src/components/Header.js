import React from 'react'
import {NavLink} from 'react-router-dom'
import {connect} from 'react-redux'
import {startLogout} from '../actions/auth'

export const Header = ({startLogout}) => (
  <header>
    <h1>Spend Track</h1>
    <NavLink to='/dashborad' activeClassName='is-active'>Home</NavLink>
    <NavLink to='/create' activeClassName='is-active'>Create Spend</NavLink>
    <button onClick={startLogout}>Logout</button>
  </header>
)
const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
})

export default connect(undefined, mapDispatchToProps)(Header)
