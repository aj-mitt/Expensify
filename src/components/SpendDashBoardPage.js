import React from 'react'
import SpendingList from './SpendingList'
import SpendingListFilters from './SpendingListFilters'
import SpendingSummary from './SpendingSummary'

const SpendDashboardPage = () => (
  <div>
    <SpendingSummary />
    <SpendingListFilters />
    <SpendingList />
  </div>

)

export default SpendDashboardPage
