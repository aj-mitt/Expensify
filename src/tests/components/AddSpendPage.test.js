import React from 'react'
import {shallow} from 'enzyme'
import {AddSpendPage} from '../../components/AddSpendPage'
import spending from '../fixtures/spending'

let startAddSpending, history, wrapper

beforeEach(() => {
  startAddSpending= jest.fn()
  history = {push: jest.fn()}
  wrapper = shallow(<AddSpendPage startAddSpending={startAddSpending} history={history} />)
})

test('Should render Add spend', () => {
  expect(wrapper).toMatchSnapshot()
})

test('Should handle onSubmit', () => {
  wrapper.find('SpendingForm').prop('onSubmit')(spending[1])
  expect(history.push).toHaveBeenLastCalledWith('/')
  expect(startAddSpending).toHaveBeenLastCalledWith(spending[1])
})
