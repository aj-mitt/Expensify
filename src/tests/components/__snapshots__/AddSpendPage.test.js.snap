// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Add spend 1`] = `
<div>
  <h1>
    Add Spending
  </h1>
  <SpendingForm
    onSubmit={[Function]}
  />
</div>
`;
